wallet: Write best block to disk before backup

fjahr opened the pull request:
I discovered that we don't write the best block to disk when trying to explain the behavior described here: https://github.com/bitcoin/bitcoin/pull/30455#discussion_r1719951882

In the context of that test, the behavior is confusing and I think it also shows that one of the already existing tests in `wallet_assumeutxo.py` doesn't actually test what it says. It only fails because the best block isn't written and actually, the height of the backup that is loaded is at the snapshot height during backup. So it really shouldn't fail since it's past the background sync blocks already.

I'm not sure if this is super relevant in practice though so I am first looking for concept ACKs on the `BackupWallet` code change. Either way, I think this behavior should be documented better if it is left as is and the test should be changed.

fjahr committed:
wallet: Write best block to disk before backup

This ensures that the best block is included in the backup which leads to a more consistent behavior when loading the backup.

furszy commented on src/wallet/wallet.cpp:
```diff
@@ -3406,6 +3406,13 @@ void CWallet::postInitProcess()
 
 bool CWallet::BackupWallet(const std::string& strDest) const
 {
+    if (m_chain) {
+        WalletBatch batch{GetDatabase()};
+        CBlockLocator loc{m_chain->getTipLocator()};
+        if (!loc.IsNull()) {
+            batch.WriteBestBlock(loc);
+        }
+    }
```
If the node's chain is ahead of the wallet, you would be writing a locator in the future. Which would make the wallet skip blocks during the next rescan attempt.
Should lock the wallet mutex and obtain the wallet's best locator from the wallet's last processed block (`m_last_block_processed`).

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/30678#discussion_r1722605697

> Should lock the wallet mutex and obtain the wallet's best locator from the wallet's last processed block (`m_last_block_processed`).

Left a similar suggestion https://github.com/bitcoin/bitcoin/issues/30686#issuecomment-2299674756 implementing that

furszy reviewed (commented):
See #30221. It fixes this in a more comprehensive way.
The comment I left is secondary.

achow101 commented:
> This could cause a wallet (and its backup) to have a best block locator that's unknown to the block index, if a crash happens in between the backup and the next database flush. How do we behave when loading such a wallet file?

No issues, we just rescan from the most recent fork indicated by the locator. It's the same as loading a wallet when the node isn't synced yet.

fjahr commented:
> See #30221. It fixes this in a more comprehensive way.

Ok, I didn't find it and weirdly there seems to be no conflict...

fjahr commented:
Reopening due to interest from @ryanofsky in https://github.com/bitcoin/bitcoin/issues/30686. I implemented the suggestion improvements and expanded the test a little including some more detailed comments. If this is an improvement already maybe this has a chance to improve things while the approach of #30221 is still being discussed.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -3406,6 +3406,14 @@ void CWallet::postInitProcess()
 
 bool CWallet::BackupWallet(const std::string& strDest) const
 {
+    if (m_chain) {
+        CBlockLocator loc;
+        WITH_LOCK(cs_wallet, chain().findBlock(m_last_block_processed, FoundBlock().locator(loc)));
+        if (!loc.IsNull()) {
+            WalletBatch batch(GetDatabase());
+            batch.WriteBestBlock(loc);
+        }
+    }
```
In commit "wallet: Write best block to disk before backup" (888a167446901c12960e6776613a3a1fb9789f59)

I wonder if it makes sense to put this in the CWallet::Flush method and call Flush() here. For example, this also seems like a good thing to do when unloading a wallet.

fjahr replied:
I just looked into this and from my understanding `CWallet::Flush()` delegates to the `Flush()` method of the underlying Database and that method is a [no-op in SQLite](https://github.com/bitcoin/bitcoin/blob/master/src/wallet/sqlite.h#L155). Since we are moving to SQLite and remove BDB entirely my guess is that the `Flush()` method will be removed at some point in the future as well.

ryanofsky reviewed (approved):
Code review ACK 888a167446901c12960e6776613a3a1fb9789f59. I think this would be a good change and improvement, however:

- I don't have a high level of confidence about this so would be interested in opinions from achow, furszy, and others who could point out other effects and drawbacks.
- If we are going to update the locator when backing up wallets, it seems like it might make sense to update it in other cases like when unloading wallets (https://github.com/bitcoin/bitcoin/pull/30678#discussion_r1724059212)
- Concern that this might make wallet sync less efficient in case of a node crash where the node's block index is not flushed https://github.com/bitcoin/bitcoin/pull/30678#issuecomment-2297842573 seems worth nothing somewhere, maybe in a code comment.

furszy committed:
wallet: migration, write best locator before unloading wallet

furszy reviewed (commented):
I'm not seeing any drawback from doing it prior to a backup so far.

But the `Flush()` discussion (https://github.com/bitcoin/bitcoin/pull/30678#discussion_r1724059212) made me think about the migration process. We aren't updating the best locator before unloading the wallet. So, might also want to add https://github.com/furszy/bitcoin-core/commit/5b618f1b024f16640a2a05f1e269f61b33b86577.

luke-jr commented:
Maybe we should have a custom locator with the last-change block, and last-scanned block?

ismaelsadeeq commented on test/functional/wallet_assumeutxo.py:
```diff
@@ -62,9 +62,15 @@ def run_test(self):
+        # Create a wallet that we will create a backup for later (at snapshot height)
         n0.createwallet('w')
         w = n0.get_wallet_rpc("w")
 
```
It would be beneficial to have this in a separate test method so that we can add more test cases cleanly without associating it with the assume UTXO test. For example, we can test https://github.com/furszy/bitcoin-core/commit/5b618f1b024f16640a2a05f1e269f61b33b86577 by creating a non-descriptor wallet, migrating it, and attempt to restore it.

fjahr replied:
Point taken that this is about backups and not assumeutxo primarily, so I have added another test for this in`wallet_backup.py`.

> It would be beneficial to have this in a separate test method so that we can add more test cases cleanly without associating it with the assume UTXO test.

I don't fully get this comment though. I don't know how I could make a method that would make this code reusable. The code in `wallet_assumeutxo.py` is spread across the test and pretty specific to the assumeutxo context. I don't know how that would have help me in the test I added.

> For example, we can test https://github.com/furszy/bitcoin-core/commit/5b618f1b024f16640a2a05f1e269f61b33b86577 by creating a non-descriptor wallet, migrating it, and attempt to restore it.

I don't know how to do this or rather I think it's much more complicated than you make it sound. There needs to be a way for the blocks close to the backup height to be missing in order to hit this issue. This can happen because we are using assumeutxo, which was my initial use case for this, and in the new test I added, I achieved this with pruning. But in the walletmigration context, both wallets (before and after migration) are on the same node and the backup is created as part of the process but without any actual user interaction. So to hit this case we would make the wallet migration fail after the backup is created, mine a few more blocks,  then prune the node to the height where the migration failed, and then try to use the backup from the failed migration.

Maybe I am missing something because I am not super familiar with the walletmigration code though.

ismaelsadeeq replied:
> The code in wallet_assumeutxo.py is spread across the test and pretty specific to the assumeutxo context. I don't know how that would have help me in the other test I added.

Never mind `test_pruned_wallet_backup` is even better than what I was thinking.

> So to hit this case we would make the wallet migration fail after the backup is created, mine a few more blocks, then prune the node to the height where the migration failed, and then try to use the backup from the failed migration.

I dont think it needs to fail, because the migration process will create a backup of the wallet before migrating.
which is still available after the wallet successfully finish migrating.

So we can just try restoring the backed up file created in that process in `test_pruned_wallet_backup`  
see https://github.com/ismaelsadeeq/bitcoin/commit/725b4975cd7d3f38a7b24f749c0342eeb522f0ee

ismaelsadeeq reviewed (commented):
Concept ACK, and ACK after resolving https://github.com/bitcoin/bitcoin/pull/30678#pullrequestreview-2260723558


Should also add a test for when a wallet is migrated, I left a suggested for that.

fjahr committed:
test: Add coverage for best block locator write in wallet_backup

fjahr commented:
Sorry for the long wait. I added the following changes:

- rebased to get rid of autotools here
- added commit from @furszy for the migration context as suggested here: https://github.com/bitcoin/bitcoin/pull/30678#pullrequestreview-2260723558
- added a separate test for this in `wallet_backup.py` to make sure this is covered not only in the assumutxo context but also the pruning context

fjahr commented:
> Maybe we should have a custom locator with the last-change block, and last-scanned block?

Seems like a bigger change, I will leave this to be explored in a follow-up. It should probably be weighed against the potential follow-up changes proposed in #30221.

furszy reviewed (commented):
Code review ACK 38648f4940eb13ebc858081c63e587156428107a

ismaelsadeeq commented on test/functional/wallet_backup.py:
```diff
@@ -140,6 +140,23 @@ def restore_wallet_existent_name(self):
+        self.log.info("Test loading backup on a pruned node with backup was created close to the prune height of the restoring node")
+        node = self.nodes[3]
+        self.restart_node(3, ["-prune=1", "-fastprune=1"])
+        # We need a few more blocks so we can actually get above an realistic
```
nit:
Assert that the chain height is 214, because the test assume so?
```suggestion
        # Ensure the chain tip is at height 214, because this test assume it is.
        assert_equal(node.getchaintips()[0]["height"], 214)
        # We need a few more blocks so we can actually get above an realistic
```

fjahr replied:
added

ismaelsadeeq commented on test/functional/wallet_backup.py:
```diff
@@ -140,6 +140,23 @@ def restore_wallet_existent_name(self):
         assert wallet_file.exists()
 
+    def test_pruned_wallet_backup(self):
+        self.log.info("Test loading backup on a pruned node with backup was created close to the prune height of the restoring node")
```
nit
```suggestion
        self.log.info("Test loading backup on a pruned node when the backup was created close to the prune height of the restoring node")

```

fjahr replied:
done

ismaelsadeeq reviewed (commented):
code review ACK 38648f4940eb13ebc858081c63e587156428107a

nice adding 38648f4940eb13ebc858081c63e587156428107a 👍🏾

fjahr committed:
test: Add basic balance coverage to wallet_assumeutxo.py

fjahr commented on test/functional/wallet_assumeutxo.py:
```diff
@@ -160,7 +171,7 @@ def run_test(self):
 
         self.log.info("Ensuring wallet can be restored from backup")
-        n1.restorewallet("w", "backup_w.dat")
+        n1.restorewallet("w2", "backup_w2.dat")
```
I have amended the log message and I have added basic coverage for the balances of the wallet. That required some further changes to the test as a whole since  there were no balances before due to the use of miniwallet, so I have done this in a separate commit.

fjahr commented:
Addressed feedback from @ismaelsadeeq and @furszy , fixing logs and adding more test coverage.

furszy commented on test/functional/wallet_assumeutxo.py:
```diff
@@ -172,6 +182,13 @@ def run_test(self):
+        assert_equal(n1.getbalance(), 340)
+
+        # Check balance of w wallet after node is synced
+        w = n1.get_wallet_rpc("w")
+        n1.loadwallet("w")
+        assert_equal(w.getbalance(), 34)
```
nit:  `w` would throw an error if used prior to loading the wallet. Should first load the wallet and secondly call `get_wallet_rpc`.

fjahr replied:
done

furszy commented on test/functional/wallet_assumeutxo.py:
```diff
@@ -111,7 +115,11 @@ def run_test(self):
-        self.generate(n0, nblocks=100, sync_fun=self.no_op)
+        for i in range(100):
+            if i % 3 == 0:
+                self.mini_wallet.send_to(from_node=n0, scriptPubKey=address_to_scriptpubkey(w_address), amount=1 * COIN)
+                self.mini_wallet.send_to(from_node=n0, scriptPubKey=address_to_scriptpubkey(w2_address), amount=10 * COIN)
```
nit: could call `address_to_scriptpubkey()` only once outside the for loop.

fjahr replied:
done

furszy reviewed (commented):
ACK bd8fee512301059cc03673099422c19b39f2ed65

fjahr commented:
Added improvements suggested by @furszy , thanks!

furszy reviewed (commented):
ACK f20fe33

achow101 commented:
ACK f20fe33e94c6752e5d2ed92511c0bf51a10716ee

ismaelsadeeq reviewed (commented):
post-merge code review and tested re-ACK f20fe33e94c6752e5d2ed92511c0bf51a10716ee
